### PR TITLE
feat: Add Psalm integration

### LINKED ISSUE #19: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `.psalm.xml.dist` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by alexraputa**
Closes #19

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### CODE REVIEW

**Comment by arueckauer** [comment]
@Ocramius This has been sitting around for a while. Does it need a review or can it be merged?

#AwaitingMaintainerResponse

**Comment by Ocramius** [comment]
@arueckauer working on getting this merged, thanks! :+1:

**Comment by Ocramius** [comment]
Aaah, nope, we can't merge this: it is full of BC breaks due to missing `final` in our classes :(

File: src/Listener/AbstractListener.php [hunk 1]
```diff
@@ -20,17 +20,12 @@ public function __construct(?ListenerOptions $options = null)
         $this->setOptions($options);
     }
 
-    /** @return ListenerOptions */
-    public function getOptions()
+    public function getOptions(): ListenerOptions
```
Comment (Ocramius) [inline src/Listener/AbstractListener.php]: These changes will break any existing child class implementation: we can't add native types without breaking inheritance.

For example, see https://3v4l.org/75MVt

Ref: https://github.com/laminas/laminas-ci-matrix-action/pull/126 (that patch saves me SO MUCH work in reviews!)

**Review by Ocramius — CHANGES_REQUESTED** [review]
Most of the `src/` changes need reverting here, due to them being too risky to apply.

**Comment by Ocramius** [comment]
I have a few minutes, will try adjusting myself :)

**Comment by Ocramius** [comment]
@alexraputa whoops, I rolled over your changes, sorry :S

**Comment by Ocramius** [comment]
:ship: thanks @alexraputa! Sorry for the noise and delay!

**Comment by alexraputa** [comment]
@Ocramius no problem 😉

**Pull request merged**